cmds/gzip: Add gzip cmd

bensallen opened the pull request:
New command, gzip.

- Attempts to replicate Pigz and gzip CLI behavior
- ~~Uses github.com/klauspost/pgzip for parallel Gzip support~~

TODOs:

- ~~Tests~~ - Good enough for now
- ~~STDIN pipe support~~ - DONE
- ~~Verbose/quiet flags~~ - DONE
- ~~Recursive support~~ - PUNT
- ~~Whatever rsyncable, no-name and name flags are supposed to do~~ - PUNT
- ~~Test flag support~~ - DONE
- ~~Figure out how to get `go build` to make symlinks for gunzip and gzcat~~ This isn't a thing go build can do.

Signed-off-by: Ben Allen <bensallen@me.com>

hugelgupf commented on cmds/gzip/gzip.go:
```diff
@@ -0,0 +1,436 @@
+
+func (o *options) parseArgs() error {
+
+	levels := make([]bool, 10)
```
`var levels [10]bool`

hugelgupf commented on cmds/gzip/gzip.go:
```diff
@@ -0,0 +1,436 @@
+}
+
+func (o *options) parseArgs() error {
+
```
super nit: remove empty line at beginning of each function?

hugelgupf commented on cmds/gzip/gzip.go:
```diff
@@ -0,0 +1,436 @@
+		"test",
+		"t",
+		false,
+		"Test the integrity of the compressed input [NOT IMPLEMENTED]")
```
Let's not include options that aren't implemented?

bensallen replied:
Another option is we hide the flags from usage output until their implemented. Only reason I can think of to keep them as valid flags and hide them is if we need the CLI compatibility somewhere.

bensallen replied:
Leaning towards just commenting out the unimplemented flags and option struct attributes for now though.

hugelgupf commented on cmds/gzip/gzip.go:
```diff
@@ -0,0 +1,436 @@
+		if o, err = os.Create(f.outputPath()); err != nil {
+			log.Fatal(err)
+		}
+	}
```
Does GNU gzip check Close return errors? Few things will actually generate an error on Close, so it'd be fine to just defer the Close here and above.

hugelgupf replied:
Especially because the way it currently is, you don't close on errors.

rminnich commented on cmds/gzip/gzip.go:
```diff
@@ -0,0 +1,436 @@
+		}
+	}
+
+	if err := i.Close(); err != nil {
```
yeah this is kinda nasty. you exit here with that file open and if the close errored we'll never know.

Also, I've been burned by deferring closes, and there is a good writeup somewhere on why deferring close is a bad idea anyway. https://joeshaw.org/dont-defer-close-on-writable-files/

if you defer close and aren't careful you can get EMFILE in weird ways. But for sure you need to test the close status of the writeable file.

bensallen commented:
Changes:

- Split vendoring into a second commit
- Cleaned-up / added various Close(), I think they're in good shape now.
- *Don't* call Close() if output is os.Stdout
- Implemented the verbose option, prints `<input file path> to <output file path>`
- Print errors to os.Stderr, especially important in an application that is expected to output data to os.Stdout.
- Including file path in error messages that was missing it.

rminnich commented on cmds/gzip/gzip.go:
```diff
@@ -0,0 +1,454 @@
+		false,
+		"Display license")
+
+	/*
```
I'm not a big fan of commented out code, why not just leave a TODO here with 
// TODO: implement list and name here
// no-name makes no sense to me , it's just the opposite of name right? Why need two?

bensallen replied:
// TODOs are a reasonable approach.

It appears --name is default on compression,

> When compressing, always save the original file name and time stamp; this is the default. When decompressing, restore the original file name and time stamp if present. This option is useful on systems which have a limit on file name length or when the time stamp has been lost after a file transfer.


and --no-name is default behavior on decompression,

> When compressing, do not save the original file name and timestamp by default. (The original name is always saved if the name had to be truncated.) When decompressing, do not restore the original file name if present (remove only the gzip suffix from the compressed file name) and do not restore the original time stamp if present (copy it from the compressed file). This option is the default when decompressing.

Both options are needed to counteract the defaults it appears. That being said, looking through klauspost/pgzip, I don't see a straight forward way to implement either of these options to negate the default behaviors noted above. I also don't actually see the use case of negating either of the default behaviors.

rminnich commented on cmds/gzip/gzip.go:
```diff
@@ -0,0 +1,454 @@
+			false,
+			"Process the contents of all subdirectories [NOT IMPLEMENTED]")
+
+		pflag.BoolVarP(&o.rsyncable,
```
I don't think this makes sense as an option. GNU is a bit option crazy.

bensallen replied:
Appears rsyncable's purpose is to make changes between gzip versions of the same / or slightly updated file into predictable blocks so Rsync only has to copy the changed blocks. Neither klauspost/pgzip or compress/gzip implement this, so it's unlikely we should bother with it.

rminnich commented on cmds/gzip/gzip.go:
```diff
@@ -0,0 +1,454 @@
+		".gz",
+		"Specify suffix for compression")
+
+	/*
```
// TODO

rminnich commented on cmds/gzip/gzip.go:
```diff
@@ -0,0 +1,454 @@
+		false,
+		"Write all processed output to stdout (won't delete)")
+
+	pflag.StringVarP(&o.suffix,
```
I don't like this style of 6 lines of code where one would do. These options would be far more readable (to me) were they done as one line. What's the reason to space them out this way?

rminnich replied:
Turns out this is what people recommend, splitting across lines? But if so you should probably tag the initializers ...

bensallen replied:
You're right. Looks better as single lines.

bensallen commented:
Should we support decompressing suffixes like .tgz even when the suffix is specified as .gz, like GNU gzip? In other words, drop the period of the specified suffix when checking that the suffix on the input file for decompression is correct? Since the suffix option includes the period, to me this feels a bit hacky, but I could see doing it for compatibility.

rminnich commented:
It's a very fine line. Do we implement all this stuff even when it's questionable? 

I'd rather implement just enough to make a reasonable program, and see if that will support our needs.

hugelgupf commented:
I would agree with Ron here.

We should just commit the basics for now, and add things later as needed.

It's easy enough to add the remaining flags when they're needed, or detection for .tgz. Otherwise it's just dead code from the beginning (or dead commented out code). I'd rather force the person implementing this later to explicitly add the flag. The more realistic case is we move on to another command because we have everything we mostly need for gzip and that commented out code will stay dead forever. Or we switch to another flags package and the converter doesn't convert the commented out code and then it's dead non-functional code.

Other than that this is great stuff!

hugelgupf commented:
Ben, does gnu gzip just allow .tgz separately, or does it actually just check for a gz suffix?

bensallen commented:
@hugelgupf Appears gzip is special casing .tgz in addition to the default .gz. It also magically knows the output suffix is .tar when special input suffix is .tgz.

$ gzip -v -d pigz-2.3.4.targz
gzip: pigz-2.3.4.targz: unknown suffix -- ignored
$ gzip -v -d pigz-2.3.4.tgz
pigz-2.3.4.tgz:	 73.8% -- replaced with pigz-2.3.4.tar

Also interestingly pigz doesn't support specifying alternate suffix on decompression, only on compression. Gzip does support alternate suffix on both compression and decompression.

I'm happy with leaving this special casing of .tgz out.

rminnich commented:
since you're using pflag ... do any packages you import use flag and what happens if they do? Will your flags and their flags work together?

bensallen commented:
@rminnich It doesn't appear any of gzip's imports use the flag package currently. I don't expect there to be any issues if for some reason it did, as I'm importing pflag as `pflag` and not renaming it on import to `flag`.

hugelgupf commented:
Ron, are we likely to use gzip in bb mode? (I could see wgeting a gzipped kernel and initramfs to execute...) I think in that case we may want to stick with regular flag until bb mode can be fixed up a little better.

(For context, bb mode currently rewrites package sources, and specifically looks for the flag package at the moment. There are a number of problems with this approach, but until we have some time to fix it, we may have to work around it.)

rminnich commented:
chris, this is a good question in general. What do we want in busybox? At what point should we cut over to dynamic? there is a speed/size tradeoff here. But busybox also affects boot speed and that's a consideration. Here's a number: moving from busybox to dynamic on nichrome adds 23 seconds to boot time, 7 seconds goes to 30 or so. I think that compromise is worth it, on a general purpose laptop that may be up for days, but on an embedded system, it's almost certainly not acceptable. 

BTW this is my concern with making our tools too fancy, e.g. ls. You definitely want ls in busybox. How much bigger does the image get when we add the human package? 

As to your question, gzip could be a real performance win if we can wget a gzip'ed thing and uncompress it, since memory bandwidth is almost always far higher than network bandwidth. This might matter for ARM. OTOH, how big are kernels likely to be? On the OCP, on 100 Mbit, the wget feels instantaneous. Not sure we need gzip, since the kernel has it anyway ...

how big is gzip now? How much does it grow the busybox initramfs? Maybe we should work that out before we try to decide.

hugelgupf commented on pkg/gzip/error.go:
```diff
@@ -0,0 +1,52 @@
+func ErrorHandler(err error, stdout io.Writer, stderr io.Writer) int {
+	switch err.(type) {
+	case *appError:
+		switch err.(*appError).level {
```
Rather than switching on the level here, appError.Error should just return the right error string. Then, you need not have the ErrorHandler at all, which seems like a little too much for this

bensallen replied:
Removed all of error.go, and simplified error handling across the app. The only thing that I didn't carry over is exit status 0 on gzip -h, due to where the help option is parsed.

bensallen commented on pkg/gzip/options.go:
```diff
@@ -0,0 +1,145 @@
+	Suffix     string
+}
+
+func (o *Options) ParseArgs() error {
```
Done.

bensallen commented on pkg/gzip/options.go:
```diff
@@ -0,0 +1,145 @@
+}
+
+// parseLevels loops through a [10]bool and returns the index of the element
+// thats true. If more than one element is true return an error. If no
```
Done

bensallen commented on pkg/gzip/options.go:
```diff
@@ -0,0 +1,145 @@
+// parseLevels loops through a [10]bool and returns the index of the element
+// thats true. If more than one element is true return an error. If no
+// element is true, return the constant pgzip.DefaultCompression (-1).
+func parseLevels(Levels [10]bool) (int, error) {
```
Search and release fail... done.

bensallen commented on pkg/gzip/options.go:
```diff
@@ -0,0 +1,145 @@
+// thats true. If more than one element is true return an error. If no
+// element is true, return the constant pgzip.DefaultCompression (-1).
+func parseLevels(Levels [10]bool) (int, error) {
+	var Level int
```
Done.

bensallen commented:
I believe this is in reasonable enough shape to be merged. Anyone want to do a finally look through / review?

hugelgupf commented on cmds/gzip/gzip.go:
```diff
@@ -0,0 +1,77 @@
+	"github.com/u-root/u-root/pkg/gzip"
+)
+
+var cmdLine = flag.CommandLine
```
Did you try this with bb mode? I don't think bb mode handles this correctly right now.

bensallen replied:
Yep works great, booted https://125-109016190-gh.circle-artifacts.com/0/bb_initramfs.linux_amd64.cpio.lzma from https://circleci.com/gh/bensallen/u-root/125#artifacts/containers/0:

```
% gzip -h

Usage of gzip:
  -1	Compression Level 1
  -2	Compression Level 2
  -3	Compression Level 3
  -4	Compression Level 4
  -5	Compression Level 5
  -6	Compression Level 6
  -7	Compression Level 7
  -8	Compression Level 8
  -9	Compression Level 9
  -S string
    	Specify suffix for compression (default ".gz")
  -b int
    	Set compression block size in KiB (default 128)
  -c	Write all processed output to stdout (won't delete)
  -d	Decompress the compressed input
  -f	Force overwrite of output file and compress links
  -h	Display a help screen and quit
  -k	Do not delete original file after processing
  -p int
    	Allow up to n compression threads (default 1)
  -q	Print no messages, even on error
  -t	Test the integrity of the compressed input
  -v	Produce more verbose output
wait: exit status 2
%
```

hugelgupf replied:
huh. cool!
